libs/iplant-clojure-commons/src/clojure_commons/exception.clj

```diff
@@ -26,102 +23,115 @@
    a ':errors' subkey in the response ':body'"
   [error-handler error-code]
   (fn [error error-type request]
-    (let [response (error-handler error error-type request)
+    (let [hndlr-resp (error-handler error error-type request)
           exception {:error_code error-code
-                     :reason (get-in response [:body :errors])}]
-      (header/content-type
-        (embed-error-info error
-                          exception
-                          (assoc response
-                            :body (cheshire/encode exception)))
-        "application/json; charset=utf-8"))))
+                     :reason (get-in hndlr-resp [:body :errors])}
+          response (header/content-type
+                     (assoc hndlr-resp :body (cheshire/encode exception))
+                     "application/json; charset=utf-8")]
+      (smw/log-response-with-exception error
+                                       request
+                                       response
+                                       exception)
+      response)))
+
+(defn default-error-handler
+  "Produces an error handler which generates exceptions with the
+  given error-code and HTTP response function."
+  [response-fn error-code]
+  (fn [error error-type request]
+    (let [exception (merge {:error_code error-code
+                            :reason (:error error-type)}
+                           (dissoc error-type :type :error))
+          response (response-fn (cheshire/encode exception))]
+      (smw/log-response-with-exception error
+                                       request
+                                       response
+                                       exception)
+      response)))
 
 (defn authentication-not-found-handler
-  [error data _]
-  (resp/unauthorized
-    (let [exception {:error_code ec/ERR_NOT_AUTHORIZED
-                     :reason (:error data)}]
-      (header/header
-        (embed-error-info error
-                          exception
-                          (resp/unauthorized (cheshire/encode exception)))
-        "WWW-Authenticate"
-        "Custom"))))
+  [error data request]
+  (let [exception {:error_code ec/ERR_NOT_AUTHORIZED
+                     :reason (:error data)}
```

Comment:
This is trivial, but this code is repeated in a few places:

```clojure
{:error_code some-error-code
 :reason        some-reason}
```

It might be worthwhile to put that into a function. Something like:

```clojure
(defn- exception-with-reason [error-code reason]
  {:error_code error-code
   :reason reason})
```

might work well. I often do that when I'm building maps in a `let` binding just to get rid of multi-line bindings too.